---
title: 'Single-Cell Workshop: Cynthia Sandor Data'
author: "Davis McCarthy, EMBL-EBI"
date: "9 September 2016"
output: 
    html_document:
        toc: true
        toc_float: true
        highlight: tango
        number_sections: true
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, message = FALSE, 
                      warning = FALSE, fig.align = "center")
suppressPackageStartupMessages(library(scater))
par(mar = c(3, 1, 1, 3))
```

# Single-Cell Data Surgery: Cynthia Sandor's Data {data-background="EMBL-EBI_Keynote_template_title_slide_bg.jpg"} 

## Aims for this dataset 

QC the data appropriately and then find genes differentially expressed between cells with different genotypes.


## Tasks for this dataset 

1. Load data then QC cells and genes
2. Normalise data
3. Investigate and correct batch effects
4. Visualise data [throughout]
5. Explore methods for differential expression (DE) analysis


# QC

## Task: QC cells and genes 

* Load the data into an SCESet object
* QC cells using appropriate functions in `scater`
* Filter out genes with very low average expression

## Load data into SCESet 

```{r load-data}

```
Expression values are log2(cpm + 1).

## Get gene annotations 

```{r get-feature-annos}

```

## View cell metadata 

```{r view-pdata}

```


```{r overview-plot, fig.height=5.5}

```

## Calculate QC metrics 

```{r calc-qc-metrics}

```



```{r plot-pheno-detected-genes-complexity, fig.height = 5}

```



```{r plot-pheno-detected-genes-complexity-2, fig.height = 5}

```



```{r plot-pheno-detected-genes-total-counts, fig.height = 5}

```



```{r plot-pca-outlier-detect, results = 'hide', message=FALSE, fig.height=4.5}

```



```{r plot-pca-exprs, results = 'hide', message=FALSE, fig.height=4.5}

```



```{r plot-pca-exprs-2, results = 'hide', message=FALSE, fig.height=4.5}

```



```{r plot-pca-exprs-3, results = 'hide', message=FALSE, fig.height=4.5}

```


## Filter cells

```{r filter-cells}

```



```{r plot-pca-exprs-post-filt, results = 'hide', message=FALSE, fig.height=4.5}

```



```{r plotqc-pcs, fig.height = 5}

```



```{r plotqc-pcs-2, fig.height = 5}

```


## PlotQC: most highly expressed genes 

```{r plotqc-highest-expression, fig.height=4}

```

## PlotQC: expression frequency against mean 

```{r plotqc-exprs-vs-mean, fig.height=4}

```

## Filter genes

```{r filter-genes}

```


# Normalise expression data

## Task: Scaling normalisation

* Use the methods from the `scran` package to apply scaling normalisation to the data.
* Look at the functions `isSpike`, `computeSumFactors`, `computeSpikeFactors` and `normalise`

## Scaling normalisation

```{r normalise-scran}

```



```{r plot-pca-exprs-norm, results = 'hide', message=FALSE, fig.height=4.5, fig.width=11}

```


## Important explanatory variables

```{r plotqc-expl-var, fig.width=8}

```



# Batch effect

## Batch effect

The PCA plots above show that scaling normalisation does not remove batch effects (this is beyond its scope).

Options for dealing with batch effects (in order of preference):

1. Design the experiment to avoid them
2. Model them explicitly
3. Adjust/correct data to remove or ameliorate them

* Here we will address (3)

## Correcting for batch effects

Options:

* Using a method to identify and remove hidden/latent factors from the data. I have had good results with the `RUVs` method from the `RUVSeq` package.
* Regressing out known factors [shown here].

## Task: Regress out Lab effect

* Use `scater`'s `normaliseExprs()` function to regress out the Lab effect.
* Compare results of this adjustment to previous normalisation results.


## Regressing out known factors

```{r regress-factors-norm}

```



```{r plot-pca-regress-factors-norm, fig.height = 4.5, fig.width=11}

```



```{r plot-tsne-regress-factors-norm, fig.height = 4.5}

```


# Differential Expression Analysis

## Do we need to filter genes/libraries beore DE analysis?

* Yes
* Filter libraries we believe to be problematic, as done in our QC previously
* We can increase power to detect DE genes (reducing multiple-testing burden) by filtering genes, **BUT**...
* We have to be careful with gene filtering: must not filter on gene attributes associated with DE testing itself, or we will bias results and p-values/FDR will no longer be accurate
* Generally safe and useful approach is to filter out genes with low expression in a way that is blind to which group/experimental condition/etc they belong to


## Tools for DE analysis

* `edgeR` [Bioc]: use `trend.method="none"` argument in `estimateDisp` call
* `scde` [Bioc]
* `M3Drop` [Bioc]
* `MAST` [devtools::install_github("RGLab/MAST@summarizedExpt"")]
* `BPSC` [devtools::install_github("nghiavtr/BPSC")]
* `scDD` [devtools::install_github("kdkorthauer/scDD")]

## What's the best method for DE?

**It depends...**

## Our methodological desires...

We To find genes DE between PSEN1 and control cells accounting for Lab effects (at least). So we want the following in a DE method/tool:

1. can handle general experimental designs
2. accounts for characteristics of single-cell expression data (e.g. bimodal distributions)
3. is fast

(a good tool should also be reliable, easily available, and well supported)

**Choose two.**

## Pros/cons

* `edgeR`: general designs, fast, reliable | negative binomial distribution (unimodal)
* `MAST`: general designs, single-cell model | fast, reliable(?)
* `BPSC`: general designs, single-cell model | fast(?), reliable(?), 
* `scde`: single-cell model | fast, reliable(?), general designs
* `scDD`: single-cell model (non-parametric) | fast(?), reliable(?), general designs(?)

## Interpretation of results

* Specific threshold on p-value and/or fold-change
* Comparison with the results of bulk experiments


## Tasks: DE analysis with edgeR

* Fit a model in edgeR with both Lab and Genotype effects
* Find genes that are DE between different genotypes

Tips:

* Use the `convertTo` function in `scran` to convert an SCESet to a DGEList for `edgeR` analysis. (This retains necessary metadata, including normalisation factors previously computed.)


## edgeR estimate dispersion

```{r edgeR-de-estimate-disp}
library(edgeR)

```

## edgeR plot biological CV

```{r plot-bcv, fig.height=5}

```

## edgeR fit quasi-likelihood GLM

```{r edgeR-fit}

```


## edgeR MA plot

```{r plot-smear, fig.height=4.5}

```

## Expression for top edgeR DE genes

```{r top-de-genes-edgeR}

```


## M3Drop - Michaelis-Menten modeling of dropouts

```{r m3drop, fig.height=4.5}
library("M3Drop")
norm_counts <- t(t(counts(sce)) / sizeFactors(sce))
m3_fit <- M3DropDropoutModels(norm_counts)
```

## M3Drop - DE genes

```{r m3drop-de, fig.height=5}
m3_de <- M3DropDifferentialExpression(norm_counts, 
			mt_method = "fdr", mt_threshold = 0.01)
```


## M3Drop - heatmap (Genotype)

```{r m3drop-heatmap-genotype, fig.height=5}
heat_out <- M3DropExpressionHeatmap(m3_de$Gene, norm_counts, 
			cell_labels = sce$Genotype)
```

## M3Drop - heatmap (Lab)

```{r m3drop-heatmap-lab, fig.height=5}
heat_out <- M3DropExpressionHeatmap(m3_de$Gene, norm_counts, 
			cell_labels = sce$Lab)
```

Beware batch effects!



## scDD

In principle, the below should work...but it did not for me. 

As it is not on CRAN or Bioconductor, there is no established forum for asking questions and obtaining advice, and (it would appear here) that it is not as reliable as it should be.

```{r scdd, eval=FALSE}
eset <- ExpressionSet(assayData(sce))
exprs(eset) <- get_exprs(sce, "exprs")
eset$condition <- as.numeric(as.factor(sce$Genotype))
sdd_res <- scDD(eset, testZeroes = FALSE)
```

I'll leave it with you to explore further options for DE testing. Many good tools are available, but it's still the wild west out there.


